pallets/bonds/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+//! * It's possible to issue new bonds for bonds that are already mature.
+//!
+//! ## Redeeming of new bonds
+//! * Bonds can be both partially or fully redeemed.
+//! * The amount of the underlying asset an account receives is 1:1 to the `amount` of the bonds redeemed.
+//! * Anyone who holds the bonds is able to redeem them.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	ensure,
+	pallet_prelude::{DispatchResult, Get},
+	sp_runtime::{
+		traits::{AccountIdConversion, AtLeast32BitUnsigned, CheckedAdd, CheckedSub},
+		ArithmeticError, DispatchError, Permill,
+	},
+	traits::{Contains, Time},
+	PalletId,
+};
+use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+use scale_info::TypeInfo;
+use sp_core::MaxEncodedLen;
+use sp_std::vec::Vec;
+
+use hydradx_traits::{AssetKind, CreateRegistry, Registry};
+use orml_traits::{GetByKey, MultiCurrency};
+use primitives::Moment;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarks;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use codec::HasCompact;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identifier for the class of asset.
+		type AssetId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen
+			+ TypeInfo;
+
+		/// Balance type.
+		type Balance: Parameter
+			+ Member
+			+ Copy
+			+ PartialOrd
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ CheckedAdd
+			+ CheckedSub
+			+ AtLeast32BitUnsigned
+			+ MaxEncodedLen
+			+ From<u128>;
+
+		/// Multi currency mechanism.
+		type Currency: MultiCurrency<Self::AccountId, CurrencyId = Self::AssetId, Balance = Self::Balance>;
+
+		/// Asset Registry mechanism - used to register bonds in the asset registry.
+		type AssetRegistry: Registry<Self::AssetId, Vec<u8>, Self::Balance, DispatchError>
+			+ CreateRegistry<Self::AssetId, Self::Balance, Error = DispatchError>;
+
+		/// Provider for existential deposits of assets.
+		type ExistentialDeposits: GetByKey<Self::AssetId, Self::Balance>;
+
+		/// Provider for the current timestamp.
+		type TimestampProvider: Time<Moment = Moment>;
+
+		/// The pallet id, used for deriving its sovereign account ID.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// The origin which can issue new bonds.
+		type IssueOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Self::AccountId>;
+
+		/// Asset types that are not permitted to be used as underlying assets.
+		type AssetTypeBlacklist: Contains<AssetKind>;
```

Comment:
flipping this to allowlist would be probably safer when we lan to indroduce new asset types